File: src/vicon_display_py/vicon_display_py/vicon_display_node.py
import rclpy
from rclpy.node import Node
from geometry_msgs.msg import PoseStamped
from rclpy.qos import qos_profile_sensor_data
import math
import time

class ViconSubscriber(Node):
    def __init__(self):
        super().__init__('vicon_display_node')

        # Create subscription with the low-latency sensor data QoS profile
        self.subscription = self.create_subscription(
            PoseStamped,
            'vicon_pose',
            self.topic_callback,
            qos_profile_sensor_data
        )

        self.get_logger().info('Vicon position subscriber started')
        
        # Initialize timing variables for latency measurement (if needed later)
        self.last_message_time = time.perf_counter()

    def euler_from_quaternion(self, q):
        """Helper to convert a quaternion message into Euler angles (roll, pitch, yaw) in radians."""
        # Roll (x-axis rotation)
        sinr_cosp = 2.0 * (q.w * q.x + q.y * q.z)
        cosr_cosp = 1.0 - 2.0 * (q.x * q.x + q.y * q.y)
        roll = math.atan2(sinr_cosp, cosr_cosp)

        # Pitch (y-axis rotation)
        sinp = 2.0 * (q.w * q.y - q.z * q.x)
        if abs(sinp) >= 1:
            pitch = math.copysign(math.pi / 2, sinp) # use 90 degrees if out of range
        else:
            pitch = math.asin(sinp)

        # Yaw (z-axis rotation)
        siny_cosp = 2.0 * (q.w * q.z + q.x * q.y)
        cosy_cosp = 1.0 - 2.0 * (q.y * q.y + q.z * q.z)
        yaw = math.atan2(siny_cosp, cosy_cosp)

        return roll, pitch, yaw

    def topic_callback(self, msg: PoseStamped):
        # Extract position data
        x = msg.pose.position.x
        y = msg.pose.position.y
        z = msg.pose.position.z

        # Convert quaternion to euler angles
        roll_rad, pitch_rad, yaw_rad = self.euler_from_quaternion(msg.pose.orientation)

        # Convert from radians to degrees
        roll = math.degrees(roll_rad)
        pitch = math.degrees(pitch_rad)
        yaw = math.degrees(yaw_rad)

        # Print position data with minimal formatting (scaled to mm as in original C++)
        print(
            f"Position: {x * 1000:8.2f} {y * 1000:8.2f} {z * 1000:8.2f} | "
            f"Orientation (deg): Roll={roll:8.3f} Pitch={pitch:8.3f} Yaw={yaw:8.3f}"
        )

def main(args=None):
    rclpy.init(args=args)
    node = ViconSubscriber()
    
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()

if __name__ == '__main__':
    main()

File: src/vicon_display/src/vicon_display_node.cpp
#include <tf2/LinearMath/Quaternion.h>  //need for the angle/orentation

#include <chrono>
#include <geometry_msgs/msg/pose_stamped.hpp>
#include <rclcpp/rclcpp.hpp>
#include <tf2_geometry_msgs/tf2_geometry_msgs.hpp>  //need for the angle/orentation

class ViconSubscriber : public rclcpp::Node
{
   public:
    ViconSubscriber() : Node("vicon_display_node")
    {
        // Create a QoS profile optimized for low latency
        // This matches what the publisher uses
        auto qos = rclcpp::SensorDataQoS();

        // Create subscription with minimal callback overhead
        subscription_ = this->create_subscription<geometry_msgs::msg::PoseStamped>(
            "vicon_pose",
            qos,
            std::bind(
                &ViconSubscriber::topic_callback, this, std::placeholders::_1));

        RCLCPP_INFO(this->get_logger(), "Vicon position subscriber started");

        // Initialize timing variables for latency measurement
        last_message_time_ = std::chrono::high_resolution_clock::now();
    }

   private:
    rclcpp::Subscription<geometry_msgs::msg::PoseStamped>::SharedPtr subscription_;
    std::chrono::high_resolution_clock::time_point last_message_time_;

    void topic_callback(const geometry_msgs::msg::PoseStamped::SharedPtr msg)
    {
        // Get current time for latency calculation
        auto current_time = std::chrono::high_resolution_clock::now();

        // Extract position data - this is the fastest way to access it
        double x = msg->pose.position.x;
        double y = msg->pose.position.y;
        double z = msg->pose.position.z;

        // get the angle for message
        tf2::Quaternion quat;
        tf2::fromMsg(msg->pose.orientation, quat);

        // convert from quat to eular angles
        double roll, pitch, yaw;
        tf2::Matrix3x3(quat).getRPY(roll, pitch, yaw);

        // convert for radian to degeee
        roll = roll * 180 / M_PI;
        pitch = pitch * 180 / M_PI;
        yaw = yaw * 180 / M_PI;

        // Print position data with minimal formatting
        // Using printf is faster than std::cout for simple output
        printf(
            "Position: %8.2f %8.2f %8.2f | Orientation (deg): Roll=%8.3f "
            "Pitch=%8.3f Yaw=%8.3f\n",
            x * 1000,
            y * 1000,
            z * 1000,
            roll,
            pitch,
            yaw);
    }
};

int main(int argc, char *argv[])
{
    rclcpp::init(argc, argv);

    // Create the subscriber node
    auto node = std::make_shared<ViconSubscriber>();

    // Spin the node - this keeps it running and processing callbacks
    rclcpp::spin(node);

    rclcpp::shutdown();
    return 0;
}